File: WA_StoreControl/Utilidades/SearchViewModel.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Web.Helpers;

namespace WA_StoreControl.Utilidades
{
    public abstract class SearchViewModel
    {
        #region Constructors
        public SearchViewModel()
        {
            Page = 1;
            SortDirection = SortDirection.Ascending;
            RecordsPerPage = 10;
            SearchType = SearchType.Quick;
            SortColumns = new Dictionary<string, string>();
            ExcludedIds = new List<int>();
        }

        public SearchViewModel(int recordsPerPage, string sortColumn, Dictionary<string, string> sortColumns)
        {
            Page = 1;
            SortDirection = SortDirection.Ascending;
            SearchType = SearchType.Quick;
            RecordsPerPage = recordsPerPage;
            SortColumn = sortColumn;
            SortColumns = sortColumns;
            ExcludedIds = new List<int>();
        }

        #endregion

        #region Properties

        /// <summary>
        /// Página actual solicitada.
        /// </summary>
        public int Page { get; set; }

        /// <summary>
        /// Nombre de la columna por la cual se deben ordenar los datos que se presentaran en la lista.
        /// </summary>
        [Display(Name = "Ordenar por")]
        public string SortColumn { get; set; }

        /// <summary>
        /// Diccionario que contiene los pares clave-valor de las columnas por las que es posible ordenar 
        /// los resultados de búsqueda.
        /// </summary>
        public Dictionary<string, string> SortColumns { get; set; }


        /// <summary>
        /// Dirección de ordenamiento de los datos. Asc o Desc.
        /// </summary>
        [Display(Name = "Dirección")]
        public SortDirection SortDirection { get; set; }


        /// <summary>
        /// Total número de registros que se recuperaron
        /// </summary>
        public int TotalRecords { get; set; }

        /// <summary>
        /// Número de registros por página
        /// </summary>
        [Display(Name = "Total/Página")]
        [Range(minimum: 1, maximum: 50, ErrorMessage = "El valor del campo {0} debe estar entre {1} y {2}.")]
        public int RecordsPerPage { get; set; }

        /// <summary>
        /// Número total de páginas obtenidos en la búsqueda
        /// </summary>
        [Display(Name = "Total páginas")]
        public int TotalPages { get; set; }


        /// <summary>
        /// Cadena de texto utilizada para filtrar los datos en modo "Búsqueda rápida"
        /// </summary>
        [Display(Name = "Palabras clave")]
        public string SearchString { get; set; }


        public SearchType SearchType { get; set; }


        public bool Paginate { get; set; }


        public List<int> ExcludedIds { get; set; }


        #endregion
    }

    public enum SearchType
    {
        Quick = 0,
        Advanced = 1
    }
}

File: WA_StoreControl/Scripts/ComponentesKO/Shared/Knockout/SearchViewModel.js
class SearchViewModel {
    constructor(data) {
        var self = this;
        data = data || {};
        self.SearchString = ko.observable(data.SearchString || "");
        self.SearchType = ko.observable(data.SearchType || 0);
        self.Page = ko.observable(data.page || 1);
        self.SortColumn = ko.observable(data.SortColumn || "Id");
        self.SortDirection = ko.observable(data.SortDirection || "Asc");
        self.RecordsPerPage = ko.observable(data.RecordsPerPage || 10);
        self.TotalRecords = ko.observable(data.TotalRecords || 0);
        self.TotalPages = ko.observable(data.TotalPages || 0);
        self.Paginate = ko.observable(typeof data.Paginate === "boolean" ? data.Paginate : true);
        self.ExcludedIds = ko.observableArray(data.ExcludedIds || []); //Id de los registros agregados a excluir en una búsqueda...
    }
    ResetDefault(except) {
        this.Page(except.includes("Page") ? this.Page() : 1);
        this.SortColumn(except.includes("SortColumn") ? this.SortColumn() : "Id");
        this.SortDirection(except.includes("SortDirection") ? this.SortDirection() : "Asc");
        this.RecordsPerPage(except.includes("RecordsPerPage") ? this.RecordsPerPage() : 10);
        this.TotalRecords(except.includes("TotalRecords") ? this.TotalRecords() : 0);
        this.TotalPages(except.includes("TotalPages") ? this.TotalPages() : 0);
        this.Paginate(except.includes("Paginate") ? this.Paginate() : true);
        this.ExcludedIds(except.includes("ExcludedIds") ? this.ExcludedIds() : []);
    }
};